Added support for tools.cmake.cmaketoolchain:extra_variables

perseoGI opened the pull request:
Changelog: Feature: add support for setting `tools.cmake.cmaketoolchain:extra_variables`
Docs: https://github.com/conan-io/docs/pull/3719
Closes: https://github.com/conan-io/conan/issues/16222

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop2/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

perseoGI committed:
Added support for tools.cmake.cmaketoolchain:extra_variables (#16222)

perseoGI commented:
**UPDATE**
https://docs.conan.io/2/reference/config_files/global_conf.html#configuration-data-operator
Merge dictionaries is already implemented and can be achieved from `CLI` and `profile` with `*=` operator

_____________________________________________________

@jcar87 wonder what would happen if there is a collision in defining `tools.cmake.cmaketoolchain:extra_variables` from different places. Will dictionaries merge? If there is a collision in a key in the dict, which will be the preference order?
E.g.

`profile`
```yaml
[settings]
...
[conf]
tools.cmake.cmaketoolchain:extra_variables={'A':'hello'}
```
`CLI`
```sh
conan install . --profile=profile -c tools.cmake.cmaketoolchain:extra_variables="{'B': 'world'}"
```
`conanfile.py`
```py
    def generate(self):
        deps = CMakeDeps(self)
        deps.generate()
        tc = CMakeToolchain(self)
        tc.blocks["generic_system"].values["extra_variables"] = {'C': 'good day'}
        tc.generate()
```
This may be the desired behaviour: 
```cmake
set(A "hello")
set(B "world")
set(C "good day")
```
But actually, `tc.blocks` replace all the content of the dict. This can be addressed by indexing with the new key:
```py
tc.blocks["generic_system"].values["extra_variables"]["C"] = "good day"
```
Also, from python things are easier, if we want to merge the whole dictionary from `CLI` or `profile` with the `extra_variables` defined in the recipe:
```py 
tc.blocks["generic_system"].values["extra_variables"] = tc.blocks["generic_system"].values["extra_variables"] | {'foo': 'bar'}
```
This will generate:
```cmake
set(B "world")
set(C "good day")
```
But still have the problem of preference order: `CLI` will override the whole dictionary defined in the `profile` section.
Do we want a merge mechanism?

franramirez688 commented:
> @jcar87 wonder what would happen if there is a collision in defining `tools.cmake.cmaketoolchain:extra_variables` from different places. Will dictionaries merge? If there is a collision in a key in the dict, which will be the preference order? E.g.

To clarify the merge behavior in the case of Python dicts, it's already explained in [this section of the Conan docs](https://docs.conan.io/2/reference/config_files/global_conf.html#configuration-data-operators).
In summary, you should use `-c "tools.cmake.cmaketoolchain:extra_variables*="{'B': 'world'}"` to update the conf already defined in the profile.

memsharded reviewed (commented):
Looks good.

The only thing I'd consider, but this is something we need to discuss with @jcar87 too, is if we wanted to move this much later in the toolchain, so it can be used to overwrite other existing variables (see for example https://github.com/conan-io/conan/issues/16240). If we do that, in the same spirit we would like to have a ``user_toolchain_append`` or similar that injects user files at the end of the toolchain, allowing users to overwrite other possible things that the toolchain is defining and doesn't have another way to fix them.

perseoGI committed:
Moved extra_variable from GenericSystemBlock to new block which will be executed later

perseoGI commented:
> Looks good.
> 
> The only thing I'd consider, but this is something we need to discuss with @jcar87 too, is if we wanted to move this much later in the toolchain, so it can be used to overwrite other existing variables (see for example #16240). If we do that, in the same spirit we would like to have a `user_toolchain_append` or similar that injects user files at the end of the toolchain, allowing users to overwrite other possible things that the toolchain is defining and doesn't have another way to fix them.

I've just pushed new changes which creates a new block to handle this new field. This block can be inserted later in order to inject variables later and override possible variables.

memsharded commented:
Lets do the following:

- Lets move the extra variables just before the ``try_compile`` block. This will allow users to re-define variables
- Lets document it explicitly that this is a potential override of CMakeToolchain defined variables, users are at their own risk
- It would be good to give it a bit more power:
   - Lets make sure that it allows users to prepend/append, maybe with the value of the variable being literal ``${VAR} newvalue``, and making sure that quotes are not breaking.
   - Lets make sure that users can define also CACHE variables, as sometimes it is necessary
   - If things fit nicely in the dict string values, fine, otherwise we might need to create some additional syntax to support it.

perseoGI committed:
Moved extra variables just before the try_compile block.

This will allow users to re-define variables

jcar87 commented on conans/test/integration/toolchains/cmake/test_cmaketoolchain.py:
```diff
@@ -1575,3 +1574,33 @@ def _assert_install(out):
+        os=Windows
+        arch=x86_64
+        [conf]
+        tools.cmake.cmaketoolchain:extra_variables={'CMAKE_GENERATOR_INSTANCE': '"${GENERATOR_INSTANCE}/buildTools/"', 'FOO': '42 CACHE' }
```
I like the feature, but I'm not too sure about defining cache variables this way?  having the spaces be separators for cmake syntax  I'm not too sure.


```
set(FOO 42 CACHE)
```

IIRC, this is not valid CMake syntax:
```
CMake Error at CMakeLists.txt:5 (set):
  set given invalid arguments for CACHE mode.

```
(I think the type and dosctring are missing, see [docs](https://cmake.org/cmake/help/latest/command/set.html#set-cache-entry))

I wonder if we could have something like what CMake uses for [presets](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html):

```
      "cacheVariables": {
        "FIRST_CACHE_VARIABLE": {
          "type": "BOOL",
          "value": "OFF"
        },
        "SECOND_CACHE_VARIABLE": "ON"
      },
```

we could maybe have something like this, where we either take the value (which shouldn't allow spaces unless they are inside double quotes?) or a dictionary with the value, whether it is cache or not, and if it is cache, the other needed info (like the type)
```
tools.cmake.cmaketoolchain:extra_variables={'CMAKE_GENERATOR_INSTANCE': '"${GENERATOR_INSTANCE}/buildTools/"', 'FOO': {value:'42', cache: true, type: 'bool'}} }
```

perseoGI replied:
Alright, take a look at the new design. It is now fully working and tested. 
Now we are supporting:
```
{
'A': 'string value',
'B': 42, 
'C': 1.3,
'D': '${ENV}/test',
'DICT': {'value': '42 sense', 'cache': true, 'type': 'BOOL', 'docstring': 'test cache var'} 
}
```
Dict values includes:
* Support for pythonic boolean inputs or string inputs (in `cache` variable)
* Descriptive error if cache variable and not `type` or `docstring` defined
* Descriptive error if cache type not in `BOOL, FILEPATH, PATH, STRING, INTERNAL` list

Generating following CMake output:
```cmake
set(A 'string value')
set(B 42)
set(C 1.3)
set(D "${ENV}/test")
set(value "42 sense" CACHE BOOL "test cache var")
```

perseoGI committed:
Added extra parse layout for cmaketoolchain:extra_variables

Now it admits not only plain key:value pairs but dictionary values in
the form:
{'value': <value>, 'cache': <true/false>, 'type': <cache_type>, 'docstring': <cache docstring> }

perseoGI committed:
Fix comments

memsharded commented on conan/tools/cmake/toolchain/blocks.py:
```diff
@@ -1003,6 +1003,46 @@ def context(self):
+        elif type(value) == dict:
+            var_value = self.get_exact_type(key, value.get("value"))
+            is_cache = value.get("cache")
+            if is_cache == True or str(is_cache).lower() == 'true':
```
```suggestion
            if is_cache:
```
The value coming from the dict should already be a boolean or ``None`` if not defined. You can assert if want to make sure, rather than accept other types and convert like ``str``

perseoGI replied:
Oh, I thought it could be nice to capture both True/False and 'true'/'false'.
If we only want to accept True/False pythonic booleans, it is easier.

memsharded commented on conan/tools/cmake/toolchain/blocks.py:
```diff
@@ -1003,6 +1003,46 @@ def context(self):
+                if not var_type:
+                    raise ConanException(f'CMakeToolchain needs type defined for cache variable "{key}"')
+                if var_type not in self.CMAKE_CACHE_TYPES:
+                    possible_types = ', '.join(self.CMAKE_CACHE_TYPES)
+                    raise ConanException(f'CMakeToolchain invalid type "{var_type}" for cache variable "{key}". Possible types: {possible_types}')
```
```suggestion
                    raise ConanException(f'CMakeToolchain invalid type "{var_type}" for cache variable "{key}". Possible types: {self.CMAKE_CACHE_TYPES}')
```
Probably printing directly the list is good?

perseoGI replied:
Looks odd when formatting the error, I've fixed it a little bit

memsharded commented on conan/tools/cmake/toolchain/blocks.py:
```diff
@@ -1003,6 +1003,46 @@ def context(self):
+            return f"\"{value}\""
+        elif type(value) == int or type(value) == float:
+            return value
+        elif type(value) == dict:
```
Check with ``isinstance(value, dict)`` rather than ``type() == type``

memsharded commented on conan/tools/cmake/toolchain/blocks.py:
```diff
@@ -1003,6 +1003,46 @@ def context(self):
+    def get_exact_type(self, key, value):
+        if type(value) == str:
+            return f"\"{value}\""
+        elif type(value) == int or type(value) == float:
```
You might use ``isinstance(x, (int, float))``

perseoGI replied:
Yep, you are right!

memsharded commented on conan/tools/cmake/toolchain/blocks.py:
```diff
@@ -1003,6 +1003,46 @@ def context(self):
+            var_value = self.get_exact_type(key, value.get("value"))
+            is_cache = value.get("cache")
+            if is_cache == True or str(is_cache).lower() == 'true':
+                var_type = value.get("type")
```
Cache variables cannot use int/float? How is that modeled?

perseoGI replied:
Cache variable `TYPE` can only have the following values:
`["BOOL","FILEPATH", "PATH", "STRING", "INTERNAL"]`

memsharded commented on test/integration/toolchains/cmake/test_cmaketoolchain.py:
```diff
@@ -1575,3 +1574,53 @@ def _assert_install(out):
+                 "windows": windows_profile})
+
+    # Test passing extra_variables from pro ile
+    client.run("install . --profile:build=windows --profile:host=windows")
```
```suggestion
    client.run("install . --profile:host=windows")
```

memsharded commented on test/integration/toolchains/cmake/test_cmaketoolchain.py:
```diff
@@ -1575,3 +1574,53 @@ def _assert_install(out):
+        tools.cmake.cmaketoolchain:extra_variables={'CMAKE_GENERATOR_INSTANCE': '${GENERATOR_INSTANCE}/buildTools/', 'FOO': '42' }
+        """)
+
+    client = TestClient(path_with_spaces=False)
```
```suggestion
    client = TestClient()
```
Why? I'd say this test could work with spaces in path

perseoGI replied:
Yep, that was copy-paste from test above hehe thanks

memsharded commented on test/integration/toolchains/cmake/test_cmaketoolchain.py:
```diff
@@ -1575,3 +1574,53 @@ def _assert_install(out):
+    assert 'set(CMAKE_GENERATOR_INSTANCE "${GENERATOR_INSTANCE}/buildTools/")' in toolchain
+    assert 'set(FOO "42")' in toolchain
+
+    # Test input from command line passing dict between doble quotes
+    client.run(textwrap.dedent(r"""
+        install . -c tools.cmake.cmaketoolchain:extra_variables="{'CMAKE_GENERATOR_INSTANCE': '${GENERATOR_INSTANCE}/buildTools/', 'FOO': 42.2, 'DICT': {'value': 1}, 'CACHE_VAR': {'value': 'hello world', 'cache': 'true', 'type': 'BOOL', 'docstring': 'test variable'}}"
+    """)
+    )
+
+    toolchain = client.load("conan_toolchain.cmake")
+    assert 'set(CMAKE_GENERATOR_INSTANCE "${GENERATOR_INSTANCE}/buildTools/")' in toolchain
+    assert 'set(FOO 42.2)' in toolchain
+    assert 'set(DICT 1)' in toolchain
+    assert 'set(CACHE_VAR "hello world" CACHE BOOL "test variable")' in toolchain
```
Don't be afraid to split these into different tests (they can be part of a ``TestCmakeToolchainExtraVariables`` class)

perseoGI replied:
That is true, but in fact, they are the same test, but with different variations. I think it may be easier to have it condensed in only one function.

memsharded commented on test/integration/toolchains/cmake/test_cmaketoolchain.py:
```diff
@@ -1575,3 +1574,53 @@ def _assert_install(out):
+    """) , assert_error=True)
+    assert 'CMakeToolchain invalid type "INVALID_TYPE" for cache variable "invalid". Possible types: BOOL, FILEPATH, PATH, STRING, INTERNAL' in client.out
+
+    client.run(textwrap.dedent("""
+        install . -c tools.cmake.cmaketoolchain:extra_variables="{'invalid': {'value': 'hello world', 'cache': 'true', 'type': 'PATH'}}"
+    """) , assert_error=True)
+    assert 'CMakeToolchain needs docstring defined for cache variable "invalid"' in client.out
```
What if we allow a default ``docstring`` that reads just the name of the variable?

perseoGI replied:
Hmm we can do it, I just made a direct translation from what CMake expects.

perseoGI reviewed (commented):
Updated code

memsharded reviewed (approved):
Nice

jcar87 reviewed (approved):
LGTM!!!

peterpuppy commented:
I have test this feature with conan version 2.4.1
but i still have problem in overwrite variable in CMakeToolchain
I set in profile
```
[conf]
tools.cmake.cmaketoolchain:user_toolchain=["{{SCE_CMAKE_DIR}}/PS5.cmake"]
tools.cmake.cmaketoolchain:generator=Visual Studio 17 2022
tools.cmake.cmaketoolchain:extra_variables*={'CMAKE_GENERATOR_TOOLSET': {'value': 'PS5Clang', 'cache': True,'type': 'STRING'}}
tools.cmake.cmaketoolchain:extra_variables*={'CMAKE_GENERATOR_PLATFORM': {'value': 'Prospero', 'cache': True,'type': 'STRING'}}
tools.cmake.cmaketoolchain:extra_variables*={'CMAKE_EXE_LINKER_FLAGS_INIT': ""}
```
Here is generated toolchain file
```
set(CMAKE_GENERATOR_PLATFORM "x64" CACHE STRING "" FORCE)

message(STATUS "Conan toolchain: CMAKE_GENERATOR_TOOLSET=ClangCL")
set(CMAKE_GENERATOR_TOOLSET "ClangCL" CACHE STRING "" FORCE)

...


set(CMAKE_GENERATOR_TOOLSET "PS5Clang" CACHE STRING "CMAKE_GENERATOR_TOOLSET")
set(CMAKE_GENERATOR_PLATFORM "Prospero" CACHE STRING "CMAKE_GENERATOR_PLATFORM")
set(CMAKE_EXE_LINKER_FLAGS_INIT "")

```
CMake actually recognizes using the above parameters 
I check the sourcecode ,it seem can not inject FORCE paramter

perseoGI commented:
Hi @peterpuppy first of all thank you for your feedback!

As you have seen, injecting "FORCE" parameter is still not available. 
We will add it ASAP!

Stay tuned!
